src/examples/java/io/nats/examples/NatsJsPullIterate.java

```diff
@@ -0,0 +1,153 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io.nats.examples;
+
+import io.nats.client.*;
+import io.nats.client.impl.ConsumerConfiguration;
+
+import java.time.Duration;
+import java.util.Iterator;
+import java.util.List;
+
+import static io.nats.examples.ExampleUtils.sleep;
+
+/**
+ * This example will demonstrate a pull subscription with:
+ * - a batch size + iterate i.e. subscription.iterate(10)
+ *
+ * Usage: java NatsJsPulliterate [-s server] [-strm stream] [-sub subject] [-dur durable]
+ *   Use tls:// or opentls:// to require tls, via the Default SSLContext
+ *   Set the environment variable NATS_NKEY to use challenge response authentication by setting a file containing your private key.
+ *   Set the environment variable NATS_CREDS to use JWT/NKey authentication by setting a file containing your user creds.
+ *   Use the URL for user/pass/token authentication.
+ */
+public class NatsJsPullIterate extends NatsJsPullSubBase {
+
+    public static void main(String[] args) {
+        ExampleArgs exArgs = ExampleArgs.builder()
+                .defaultStream("iterate-ack-stream")
+                .defaultSubject("iterate-ack-subject")
+                .defaultDurable("iterate-ack-durable")
+                .build(args);
+        
+        try (Connection nc = Nats.connect(ExampleUtils.createExampleOptions(exArgs.server))) {
+
+            createStream(nc, exArgs.stream, exArgs.subject);
+
+            // Create our JetStream context to receive JetStream messages.
+            JetStream js = nc.jetStream();
+
+            // Build our consumer configuration and subscription options.
+            // make sure the ack wait is sufficient to handle the reading and processing of the batch.
+            // Durable is REQUIRED for pull based subscriptions
+            ConsumerConfiguration cc = ConsumerConfiguration.builder()
+                    .ackWait(Duration.ofMillis(2500))
+                    .build();
+
+            PullSubscribeOptions pullOptions = PullSubscribeOptions.builder()
+                    .durable(exArgs.durable) // required
+                    .configuration(cc)
+                    .build();
+
+            // 0.1 Initialize. subscription
+            // 0.2 Flush outgoing communication with/to the server, useful when app is both publishing and subscribing.
+            System.out.println("\n----------\n0. Initialize the subscription and pull.");
+            JetStreamSubscription sub = js.subscribe(exArgs.subject, pullOptions);
+            nc.flush(Duration.ofSeconds(1));
+
+            // 1. iterate, but there are no messages yet.
+            // -  Read the messages, get them all (0)
+            System.out.println("----------\n1. There are no messages yet");
+            Iterator<Message> iterator = sub.iterate(10);
+            List<Message> messages = report(iterator);
+            messages.forEach(Message::ack);
```

Comment:
Might be cool for one of these to define a lambda where users can plug in code.  Users often just copy sample code (I do all the time), adding a placeholder would be a nice touch.  e.g. something like:

```java
    messages.forEach(msg -> {
        // process message
        msg.ack();
    });
```